src/main/ruby/truffleruby/core/string.rb

```diff
@@ -70,6 +70,54 @@ def byteslice(index_or_range, length = undefined)
     byteslice index, length
   end
 
+  def bytesplice(*args)
+    is_range = Primitive.is_a?(args[0], Range)
+
+    if args.size == 3
+      start = Primitive.rb_to_int(args[0])
+      start += bytesize if start < 0
+
+      len = Primitive.rb_to_int(args[1])
+      str = StringValue(args[2])
+    elsif args.size == 2
+      unless is_range
+        raise TypeError, "wrong argument type #{Primitive.class(args[0])} (expected Range)"
+      end
+
+      start, len = Primitive.range_normalized_start_length(args[0], bytesize)
+      len = 0 if len < 0
+      str = StringValue(args[1])
+    else
+      raise ArgumentError, "wrong number of arguments (given #{args.size}, expected 2..3)"
+    end
+
+    if len < 0
+      raise IndexError, "negative length #{len}"
+    end
+
+    if bytesize < start || start < 0
+      if is_range
+        raise RangeError, "#{args[0]} out of range"
+      else
+        raise IndexError, "index #{args[0]} out of string"
+      end
+    end
+
+    len = bytesize - start if len > bytesize - start
+    finish = start + len
+
+    unless Truffle::StringOperations.on_codepoint_boundary?(self, start)
```

Comment:
Should use the Primitive from the String#byteindex PR

Reply:
Sure

Reply:
Does this make sense? https://github.com/oracle/truffleruby/pull/3044/files#diff-4dcdb35dd5ecc15240f4cf70cbc64a93be1f55415477ce35a285af09483c1d26R4462-R4472

Reply:
Yes, looks good, I left some comments